### ISSUE TITLE
Full-width character block (CJK) breaks rendering

### ISSUE THREAD

**Issue by friedbunny** [mapbox/mapbox-gl-native#931]
The fontstack can't serve the [full-width latin characters](https://en.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms) that are unfortunately common in Chinese, Japanese, and Korean (CJK), which results in entire sections of the map failing to draw when they're used. This is especially problematic when using country-localized labels (i.e., `name` and not `name_en`) in a style.

EDIT: Kanji and other CJK character sets render without any problems, only this specific block fails.

Currently, attempting to use these full-width characters throws an exception when trying to load the remote glyph PBF:

```
[ERROR] ParseTile: Parsing [8/227/99] failed: [ERROR] failed to load glyphs: HTTP status code 400; 
Raw response: {"message":"Invalid Range"}; 
URL: https://api.tiles.mapbox.com/v4/fontstack/Open%20Sans%20Regular%2c%20Arial%20Unicode%20MS%20Regular/65280-65533.pbf?access_token=pk.eyJ1IjoiZnJpZWRidW5ueSIsImEiOiJUbFBmei1zIn0.lZVQjkwWlc07RS4oEQITqg
```

[Here is an example of full-width numbers on a Japanese highway on OSM.](http://www.openstreetmap.org/edit?editor=id&relation=3793529#map=16/36.4500/137.2592) The left section of the tunnel is correct (国道41号), while the right has full-width numbers (国道４１号).

Ultimately, the fontstack needs to be able to serve full-width latin characters or MBGL should replace them with regular latin, half-width characters. #930 only temporarily fixes the drawing issues by returning a common glyph range, which means the full-width characters don't get drawn. ([Mapbox-GL-JS currently has a similar behavior.](https://friedbunny.github.io/mapbox-gl-style-dark/example.html#16.35/36.45113/-222.74053))

Cheers!

**Comment by friedbunny** [comment]
BTW: To reproduce here on native, just zoom around Japan for a while — you'll notice the above error and lots of unrendered blocks, even with the stock `name_en`.

**Comment by kkaefer** [comment]
To clarify, does this only affect full-width latin characters, like the numbers used on highways, or does it affect all full-width characters? It seems like regular full-width (like 国道) are rendering just fine.

**Comment by friedbunny** [comment]
It only affects the characters and symbols in range 65280-65535 (i.e., at the end of the basic unicode range), so far as I can tell.

This ["Halfwidth and Fullwidth Forms" block](https://en.wikipedia.org/wiki/Halfwidth_and_Fullwidth_Forms_(Unicode_block)) includes full-width latin alphabet, roman numbers, basic symbols, and half-width Japanese and Korean syllabaries. The half-width Japanese and Korean aren't as common, relatively, but I'm sure somebody has tried to use them on OSM, too.

**Comment by jfirebaugh** [comment]
It looks like there are multiple issues here.

The range in the URL above -- `65280-65533` -- is indeed invalid from the fontstack API's perspective. It should be `65280-65535`. So gl-native is somehow generating invalid ranges.

But, when I feed [the correct range](https://api.tiles.mapbox.com/v4/fontstack/Open%20Sans%20Regular%2c%20Arial%20Unicode%20MS%20Regular/65280-65535.pbf?access_token=pk.eyJ1IjoiZnJpZWRidW5ueSIsImEiOiJUbFBmei1zIn0.lZVQjkwWlc07RS4oEQITqg) to the fontstack, it produces a 500 error. So there's probably a boundary condition error there too.

**Comment by friedbunny** [comment]
[mbgl/text/glyph.cpp](https://github.com/mapbox/mapbox-gl-native/blob/22f02ac0072a8e2e23a059e08b70ccc865723803/src/mbgl/text/glyph.cpp#L10) purposely truncates the range:

```cpp
// Note: this only works for the BMP
GlyphRange getGlyphRange(char32_t glyph) {
    unsigned start = (glyph/256) * 256;
    unsigned end = (start + 255);
    if (start > 65280) start = 65280;
    if (end > 65533) end = 65533;
    return { start, end };
}
```

**Comment by mikemorris** [comment]
Hah, this could very well be a typo - possible that the intended max extent should actually be `0xFFFF` or `65535` to properly contain the [Halfwidth and Fullwidth Forms](http://jrgraphix.net/r/Unicode/FF00-FFEF) and [Specials](http://jrgraphix.net/r/Unicode/FFF0-FFFF) blocks. We currently aren't making any effort to support [blocks in the `0x10000` to `0xE007F` range](http://jrgraphix.net/research/unicode_blocks.php) because of gaps, although perhaps we should?

Check out [node-fontnik](https://github.com/mapbox/node-fontnik) for the code that's actually pulling these glyphs from [Freetype](http://www.freetype.org/) for more details @friedbunny.

**Comment by mikemorris** [comment]
@jfirebaugh The 500 error is because a `65280-65535.pbf` file doesn't exist. The actual block is `65280-65533.pbf`, so the blocks are created invalid too, not just requested invalid. The fontstack API is just (correctly) catching the invalid range before it attempts to actually retrieve the block.

**Comment by mikemorris** [comment]
Oops this looks wrong https://github.com/mapbox/node-fontnik/blob/master/bin/build-glyphs#L27-L33

**Comment by friedbunny** [comment]
Thanks @mikemorris, I didn't know where fontstack came from, repo-wise. That looks extremely promising. :+1:

**Comment by jfirebaugh** [comment]
@mikemorris So the block _is_ supposed to be `65280-65535` everywhere, and both places where there's a special case for `65533` are wrong?

**Comment by mikemorris** [comment]
Yep, correct @jfirebaugh

**Comment by friedbunny** [comment]
@mikemorris Has mapbox/node-fontnik/pull/72 found its way into production yet? The [fontstack request for this glyph range](https://api.tiles.mapbox.com/v4/fontstack/Open%20Sans%20Regular%2c%20Arial%20Unicode%20MS%20Regular/65280-65535.pbf?access_token=pk.eyJ1IjoiZnJpZWRidW5ueSIsImEiOiJUbFBmei1zIn0.lZVQjkwWlc07RS4oEQITqg) still returns `ERROR 500: Internal Server Error`.

**Comment by friedbunny** [comment]
The above fontstack fix has landed in production and #946 will fix everything here once it's merged, yay!

![fixed](https://cloud.githubusercontent.com/assets/1198851/6703233/b3d08f70-ccf5-11e4-9d27-3dd10da97ab6.png)
